### DISCUSSION TITLE
Issue with "id_token" check in "oauthCallback" flow

### DISCUSSION THREAD

**Discussion by aidenfoxx** [panva/openid-client#459] [Q&A] [answered]
Hello. In https://github.com/panva/node-openid-client/commit/92ffee5c63dc31fb578c731572cf0f83a7b53f1d a check was added that would reject `oauthCallback` requests if they contain an `id_token`. This has caused an issues with the OAuth flow in Directus where `openid-client` is used.

https://github.com/directus/directus/discussions/11048

The issue is that Facebook always returns "id_token" in its OAuth flow. There is an alternate OpenID flow but it does not support the "code" authentication type.

We can't handle this nicely since we don't know if the response contains an "id_token" before triggering an `oauthCallback`, but calling `oauthCallback` triggers an exception. Obviously we could handle the exception and trigger `callback`, but it personally feels better to revert the "id_token" check as I don't believe it adds value.

I would be curious to hear what you think?

**Comment by panva** [comment, marked as answer]
Are you saying you don't know what type of authentication request you triggered? Because you either trigger an openid connect flow (regardless of what response_type) with the openid scope or you don't. Meaning on callback you either use callback or oauthCallback. You always know what you expect from the callback upfront.

Basically, if you run into the check the fix added, you're using the client wrong. The value is there, if you do trigger openid requests you ought to take advantage of it, routing those responses to oauthCallback is wrong, unintended, and removes the many checks openid connect puts in place.

**Reply by aidenfoxx** [reply]
As you mention, it could be an issue of scope, but from what I recall Facebook always returns an "id_token" no matter if openid is specified in the scope. This was a while ago and I'm unable to verify at the moment. I'll ask the reporter if they can test without "openid" in the scope.

**Reply by panva** [reply]
And the problem is? If fb always returns an openid response (which i doubt) use the right method for it. I fail to see an issue here other than a piece of software probably being configured wrong or maybe not using this library correctly.

**Reply by aidenfoxx** [reply]
That would be fine if we were only integrating against Facebook, but this is part of a generic authentication system that serves 100,000s of users, all of which could use different providers. We could add an `if (url === 'FACEBOOK_OAUTH_URL') { callback(); } else { oauthCallback() }` but that is ugly for obvious reasons.

I understand where you're coming from. You don't want people using OAuth when they should be using OpenID. But sometimes quirks like this come up, and a provider sends back something it shouldn't. 🤷 

I've verified with the OP and their scope is "email public_profile". So Facebook is always sending back an "id_token" regardless.

**Reply by aidenfoxx** [reply]
For further reference, the endpoints returning the "id_token" are specifically OAuth:

```
AUTH_FACEBOOK_AUTHORIZE_URL="https://www.facebook.com/dialog/oauth"
AUTH_FACEBOOK_ACCESS_URL="https://graph.facebook.com/oauth/access_token"
```

**Reply by panva** [reply]
Intentionally using oauthCallback for openid authorization servers is a disservice to those 100,000s users.

**Reply by aidenfoxx** [reply]
As I said, these are the Facebook OAuth endpoints from the Facebook OAuth docs, called against the `oauthCallback`. If you're gonna patronize anyone, please patronize Facebook! 😂

**Reply by panva** [reply]
I think I'm not getting my point across. I don't particularly care about Facebook's quirks. But the issue in general tells me that there's a piece of code using openid-client for a "generic authentication system" with "different providers" and it's not set up to be told "expect pure oauth" or "expect openid" and instead only uses "oauthCallback" because it's "flexible", which in itself is a problem.

Having looked at the oauth2 driver linked from the aforementioned issue in directus/directus, I can see fundamental problems in its use and consumption of oauth / openid interfaces. For instance, as mentioned above it only uses "oauthCallback()" but then happens to call "TokenSet.prototype.claims()" to parse claims out of an ID Token. Well guess what, the ID Token claims and signature was not validated as part of "oauthCallback()".

This flawed use of this library's API is exactly why the fix was put in place. To prevent such 🦶🔫

**Reply by aidenfoxx** [reply]
I understand, and agree that the `TokenSet.prototype.claims()` is exploiting a loophole to do something it shouldn't. But that exists purely as a fallback if no "userinfo" endpoint is defined. We also have a dedicated openid flow which we always encourage to be used if possible, and that does things correctly.

But in this example I've outlined we're not actually doing things incorrectly. We use the Facebook oauth endpoints, and as such you would expect we'd call `oauthCallback`? Calling the openid `callback` for what we expect to be an oauth response is also pretty weird? I'm not even sure if the response will pass the openid `callback`.

**Reply by aidenfoxx** [reply]
From my side, it would make more sense for the `oauthCallback` to accept the oauth response but exclude the "id_token" from the `tokenSet` instead of throwing an error?

**Reply by panva** [reply]
> I've verified with the OP and their scope is "email public_profile". So Facebook is always sending back an "id_token" regardless.

I have just checked with a fresh Facebook implementation, it is not until I add the `openid` scope to the authorization request that an ID token is returned by the token_endpoint. That's the behaviour as I expect it. Without the `openid` scope I get no ID Token and I can use the `oauthCallback` method.

**Reply by panva** [reply]
Also, when i add the `openid` scope I can use regular `callback()` just fine, the only caveat is that the Issuer instance needs to be constructed manually since their well-known does not list the token_endpoint.

**Reply by panva** [reply]
Bottom line, facebook did not magically start adding ID Tokens to non-openid flows, your issue's OP is somehow requesting the "openid" scope which makes `callback` the appropriate method to use instead of `oauthCallback`.

This has been quite time consuming, it has however reassured me that this validation should stay in place to avoid 🦶🔫.

**Reply by aidenfoxx** [reply]
Hm... Sincere apologies for wasting your time then! I was fairly certain I'd gotten an "id_token" back from FB in the past without "openid" in the scope, and the OP confirmed my bias. But I guess there is something wrong on their side. Never the less, I will investigate this over the weekend and I'll be sure to remove the hacky `TokenSet.prototype.claims()` fallback in the oauth provider.

Thanks for all your assistance!

**Reply by aidenfoxx** [reply]
@panva I debugged the issue and it appears both parties were correct. It is possible for Facebook to always return an "id_token" regardless of the defined "scope". This only occurs if the user has authorized the app with "openid" in their scope in the past. From that point Facebook always considers the login to be "openid".

To solve this all I had to do was revoke access in my Facebook account and then everything works as you describe.
